### PR TITLE
cjdns: added option addExtraHosts

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

This commit restores the automatic addition of cjdns ipv6 address -> hostname mapping
as used to work in nixos-16.09 and before.  It is now optional and turned off by default
because some people complained, that evaluating nixpkgs compiled cjdns
(needed for publicKey to ipv6 conversion).

###### Things done

- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/networking/cjdns.nix [hunk 1]
```diff
@@ -19,18 +19,35 @@ let
...
+      };
     };
   };
 
+  # Additional /etc/hosts entries for peers with an associated hostname
+  cjdnsExtraHosts = import (pkgs.runCommand "cjdns-hosts" {}
+    # Generate a builder that produces an output usable as a Nix string value
+    ''
+      exec >$out
+      echo \'\'
```
Comment (Mic92) [inline nixos/modules/services/networking/cjdns.nix]: What is the purpose of the extra `''` in this case?

File: nixos/modules/services/networking/cjdns.nix [hunk 2]
```diff
@@ -19,18 +19,35 @@ let
...
+      ${concatStringsSep "\n" (mapAttrsToList (k: v:
+          optionalString (v.hostname != "")
+            "echo $(${pkgs.cjdns}/bin/publictoip6 ${v.publicKey}) ${v.hostname}")
+          (cfg.ETHInterface.connectTo // cfg.UDPInterface.connectTo))}
```
Comment (Mic92) [inline nixos/modules/services/networking/cjdns.nix]: I find the following a little bit more readable:

```
optionalString (v.hostname != "") ''
  echo $(${pkgs.cjdns}/bin/publictoip6 ${v.publicKey}) ${v.hostname}
'')
```

**Comment by spacekitteh** [comment]
if it works it's cool with me

**Comment by joachifm** [comment]
To preserve history, I think it'd be better to revert 60ded3f3632d221ca3f82a52392e155517880456 and then add the conditional as a separate commit.

File: nixos/modules/services/networking/cjdns.nix (hunk 1 above)
Comment (joachifm) [inline nixos/modules/services/networking/cjdns.nix]: The generated output is intended to look like
```nix
''
  foo bar
''
```

**Comment by joachifm** [comment]
Closed in deference to #21592

**Pull request closed without merging**